### PR TITLE
NMS-10295: Minion status tracker

### PR DESCRIPTION
**Pull request by RangerRick**
This PR adds a Minion status tracker that watches for outage events and updates the "status" column in the `monitoringsystems` entry for all minions.  It also refreshes the current state from the database at a configurable interval in case any events are missed.

* JIRA: http://issues.opennms.org/browse/NMS-10295

Our [continuous integration system](http://bamboo.internal.opennms.com:8085) will test and verify your changes.

### CODE REVIEW

File: opennms-dao/src/main/java/org/opennms/netmgt/dao/hibernate/OutageDaoHibernate.java [hunk 1]
```diff
@@ -103,6 +115,66 @@ public OnmsOutage currentOutageForService(OnmsMonitoredService service) {
...
+    /** {@inheritDoc} */
+    @Override
+    public Collection<OnmsOutage> matchingLatestOutages(final ServiceSelector selector) {
+        final Set<String> matchingAddrs = m_filterDao.getIPAddressList(selector.getFilterRule()).stream().map(InetAddressUtils::str).collect(Collectors.toSet());
+        final List<Integer> outageIds = getHibernateTemplate().execute(new HibernateCallback<List<Integer>>() {
+            @Override
+            @SuppressWarnings("unchecked")
+            public List<Integer> doInHibernate(Session session) throws HibernateException, SQLException {
+                return (List<Integer>) session.createSQLQuery(
+                    "SELECT DISTINCT MAX(outageID) AS outageID\n" +
```
Comment (j-white) [inline opennms-dao/src/main/java/org/opennms/netmgt/dao/hibernate/OutageDaoHibernate.java]: It seems inefficient to have to pull all of the latest outage ids for all of the services from the database when were only actually interested in a small subset of these.

File: opennms-dao/src/main/java/org/opennms/netmgt/dao/hibernate/OutageDaoHibernate.java [hunk 2]
```diff
@@ -103,6 +115,66 @@ public OnmsOutage currentOutageForService(OnmsMonitoredService service) {
...
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Collection<OnmsOutage> matchingOutages(final ServiceSelector selector) {
+        final Set<InetAddress> matchingAddrs = new HashSet<InetAddress>(m_filterDao.getIPAddressList(selector.getFilterRule()));
+        @SuppressWarnings("unchecked")
+        final Collection<OnmsOutage> outages = (Collection<OnmsOutage>)getHibernateTemplate().findByNamedParam("FROM OnmsOutage AS o WHERE o.monitoredService.serviceType.name IN (:services)","services", selector.getServiceNames());
+        return initializeOutage(
+                                outages.parallelStream()
```
[the remark this answers is not in the document]
Comment (RangerRick) [reply opennms-dao/src/main/java/org/opennms/netmgt/dao/hibernate/OutageDaoHibernate.java]: I suppose not.  Originally this was getting every outage entry (even multiples for a single service) so I feared it was a lot bigger and would benefit from parallelism, but now that's probably not so much the case.

File: opennms-dao/src/main/java/org/opennms/netmgt/dao/hibernate/OutageDaoHibernate.java (hunk 1 above)
Comment (RangerRick) [reply opennms-dao/src/main/java/org/opennms/netmgt/dao/hibernate/OutageDaoHibernate.java]: I'll see what I can do.  Every time I tried to get fancy with this query it started doing the weird insert thing, but that's when I was doing it in HQL.

File: opennms-dao/src/main/java/org/opennms/netmgt/dao/hibernate/OutageDaoHibernate.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (RangerRick) [reply opennms-dao/src/main/java/org/opennms/netmgt/dao/hibernate/OutageDaoHibernate.java]: I guess if I'm forced to do it in SQL anyways, I could just go back to the Big Crazy SQL Statement I had from the beginning with the self-join.  That would probably be most efficient.

File: opennms-dao/src/main/java/org/opennms/netmgt/dao/hibernate/OutageDaoHibernate.java [hunk 3]
```diff
@@ -103,6 +111,76 @@ public OnmsOutage currentOutageForService(OnmsMonitoredService service) {
...
+                        .append("        node.foreignSource,\n")
+                        .append("        node.foreignId,\n")
+                        .append("        node.location\n")
+                        .append("FROM outages\n")
+                        .append("        LEFT JOIN ifServices ON outages.ifServiceId = ifServices.id\n")
+                        .append("        LEFT JOIN service ON ifServices.serviceId = service.serviceId\n")
+                        .append("        LEFT JOIN ipInterface ON ifServices.ipInterfaceId = ipInterface.id\n")
+                        .append("        LEFT JOIN node ON ipInterface.nodeId = node.nodeId\n")
+                        .append("        LEFT JOIN (\n")
+                        .append("                SELECT DISTINCT\n")
```
Comment (j-white) [inline opennms-dao/src/main/java/org/opennms/netmgt/dao/hibernate/OutageDaoHibernate.java]: This statement still ends up doing a sequential scan of the whole outages table.

I'm wondering if we really need to obtain the last outage record for resolved outages? If we only select those that are still open this query becomes much simpler and possible to do with HQL / criteria. The lack of the record implies that the outage was resolved.

**Comment by RangerRick** [comment]
@j-white does this look good to you at this point? It's gone green and has been simplified quite a bit.  I'd like to work on some of the simpler enhancements once this is merged.

**Comment by j-white** [comment]
I gave this a spin and it worked fine - the status was properly reflected as I took a Minion up and down.

I did see some warnings in the logs for unrelated outages though:
```
2018-08-21 13:18:53,436 WARN  [pool-18-thread-1] o.o.m.s.MinionStatusTracker: Unhandled 'down' outage record: ifservice=56, lost=2018-08-21T13:14:19.696-0400, regained=null
2018-08-21 13:18:53,437 WARN  [pool-18-thread-1] o.o.m.s.MinionStatusTracker: Unhandled 'down' outage record: ifservice=57, lost=2018-08-21T13:14:19.696-0400, regained=null
```

Given that these are not related to the Minion(s) we probably shouldn't be logging warnings in this context.

**Comment by RangerRick** [comment]
@j-white hrm, that means that the query `m_outageDao.newestCurrentOutages(Arrays.asList(MINION_HEARTBEAT, MINION_RPC));` is somehow getting non-minion services, so I guess I need to do more testing on that.

**Comment by RangerRick** [comment]
OK, these latest changes should fix it.

**Comment by j-white** [comment]
Confirmed that the warnings are gone now.

**Pull request merged**